src/main/java/com/commercetools/pingpong/model/Game.java

```diff
@@ -9,78 +10,107 @@ public Game(Player firstPlayer, Player secondPlayer) {
         this.playerTwo = secondPlayer;
     }
 
-    public boolean hasSomebodyWon() {
-        return (hasPlayerWonSet(playerOne, playerTwo) || hasPlayerWonSet(playerTwo, playerOne));
+    public void updateScoreOfPlayer(Player player, int actionType) {
+
+        if (actionType == 1 && !hasSomebodyWonSet()) {
+            player.addPoint();
+            decideWhoServes();
+
+        } else if (actionType == 2 && player.getSetScore() > 0) {
+            player.subPoint();
+            whoServesAfterDoubleClick();
+
+        } else if (actionType == 3) {
+            resetSetScores();
+            resetMatchScores();
+            resetServe();
+
+        }
+
+    }
+
+    public void updateMatchScoreOfPlayers() {
+        if (hasPlayerWonSet(playerOne, playerTwo) && !isMatchOver()) {
+            this.playerOne.addMatchScore();
+            resetSetScores();
+            whoBeginsServing();
+
+        } else if (hasPlayerWonSet(playerTwo, playerOne) && !isMatchOver()) {
+            this.playerTwo.addMatchScore();
+            resetSetScores();
+            whoBeginsServing();
+        }
     }
 
-    public boolean hasOneOfThePlayersTwoPointsMore(Player playerWhoIsWinning, Player playerWhoIsLosing) {
-        return (playerWhoIsWinning.getSetScore() - playerWhoIsLosing.getSetScore() == 2);
+    public void isGameOver() {
```

Comment:
We should probably split this into 2 functions:
* `public boolean canEndGame() { return this.playerOne.getMatchScore() == 2 || this.playerTwo.getMatchScore() == 2; }`
* `public void endGame() { ... }`

Then you can use those as:

```java
if (match.canEndGame()) {
  match.endGame();
}
```

Reply:
Yes, actually there is a **canEndGame** already
7f478b8676bb69b6b7de7cba6f3d42fbbb92ffdb